conformance/runner.js

```diff
@@ -0,0 +1,321 @@
+function loadTestCases() {
+  const protobufRoot = new protobufjs.Root();
+
+  protobufRoot.resolvePath = function(origin, target) {
+    if (/^google\/.*/.test(target)) {
+      target = path.join(googleProtoFiles(), target.substr('google/'.length));
+    }
+    return target;
+  };
+
+  const protoDefinition = protobufRoot.loadSync(
+    path.join(__dirname, 'test_definition.proto')
+  );
+
+  const binaryProtoData = require('fs').readFileSync(
+    path.join(__dirname, 'test_data.binprotos')
+  );
+
+  const testType = protoDefinition.lookupType('tests.Test');
+
+  const testCases = [];
+
+  for (let offset = 0; offset < binaryProtoData.length; ) {
+    const messageLength = varint.decode(binaryProtoData, offset);
+    offset += varint.encodingLength(messageLength);
+
+    let testCase = testType.decode(
+      binaryProtoData.slice(offset, offset + messageLength)
+    );
+    testCases.push(testCase);
+
+    offset = offset + messageLength;
+  }
+
+  return testCases;
+}
+
+/** Converst a JSON object into a JS Object suitable for the Node API. */
+function convertInput(json) {
+  const obj = JSON.parse(json);
+
+  function convertValue(value) {
+    if (is.object(value)) {
+      return convertObject(value);
+    } else if (is.array(value)) {
+      return convertArray(value);
+    } else if (value === 'Delete') {
+      return Firestore.FieldValue.delete();
+    } else if (value === 'ServerTimestamp') {
+      return Firestore.FieldValue.serverTimestamp();
+    }
+
+    return value;
+  }
+
+  function convertArray(arr) {
+    for (let i = 0; i < arr.length; ++i) {
+      arr[i] = convertValue(arr[i]);
+    }
+    return arr;
+  }
+
+  function convertObject(obj) {
+    for (const key in obj) {
+      if (obj.hasOwnProperty(key)) {
+        obj[key] = convertValue(obj[key]);
+      }
+    }
+    return obj;
+  }
+
+  return convertValue(obj);
+}
+
+/** Converts a CommitRequest in Proto3 JSON to Protobuf JS format. */
+function convertCommit(commitRequest) {
+  const deepCopy = JSON.parse(JSON.stringify(commitRequest));
+  for (let write of deepCopy.writes) {
+    if (write.update) {
+      write.update.fields = convert.documentFromJson(write.update.fields);
+    }
+  }
+
+  return deepCopy;
+}
+
+/** Converts a Protobuf Precondition to the expected user input. */
+function convertPrecondition(precondition) {
+  const result = JSON.parse(JSON.stringify(precondition));
+  if (result.updateTime) {
+    result.lastUpdateTime = DocumentSnapshot.toISOTime(result.updateTime);
+    delete result.updateTime;
+  }
+  return result;
+}
+
+/** List of test cases that are ignored. */
+const ignored = [
+  // Firestore Node does not omit empty writes
+  'ServerTimestamp alone',
+  'nested ServerTimestamp field',
+  'multiple ServerTimestamp fields',
+  'ServerTimestamp with dotted field',
+
+  // Test data doesn't allow update with {a:delete}
+  'Delete cannot be nested',
+
+  // Node doesn't support field masks for set().
+  'Merge with a field',
```

Comment:
I don't understand why you're ignoring these. Some of them are ordinary uses of merge. In fact, "Merge field is not a leaf" is example 5 in go/set-update.

Reply:
Node only supports `{ merge:true }`. I need to ignore any tests that have field paths set. Once I pull in the changes from https://github.com/GoogleCloudPlatform/google-cloud-common/pull/235, I can likely prune this list.

Reply:
Oh, I re-read the spec and now I get it.